### PR TITLE
Add WikiText dataset support

### PR DESCRIPTION
**Pull request by esthermichelle**
## Summary

Added built-in support for the `Salesforce/wikitext` dataset.

This includes a dedicated WikiText dataset configuration, transform, registry entry, documentation template, and tests following the existing Hugging Face dataset support pattern.

This PR addresses the WikiText dataset part of #125.

## AI usage

- [x] AI tools were used to assist this contribution

**Tool(s) and usage:**

Claude was used to help me understand the existing dataset architecture and How to implement pattern, and to clarify how to add the WikiText dataset support.

## Testing

- [x] Tests added or updated where needed
- [x] Existing tests and checks pass
- [x] Documentation updated if needed

### CODE REVIEW

**Comment by aaishwarymishra** [comment]
btw @esthermichelle do you have worked with hugging face datasets ? or have interest in machine learning?

**Comment by esthermichelle** [comment]
Yes, I have worked with hugging face datasets and I am very interested in machine learning. I've been learning and working on ML-related projects this summer

**Comment by aaishwarymishra** [comment]
Ok, I am not against the AI use, but the issue was intermediate level as it spanned over multiple files, I would suggest you try to read and understand the complete generated code yourself first and then we can continue with the PR :)

File: trainite/datasets/wikitext.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        token_ids = self.tokenizer(
+            text,
+            add_special_tokens=False,
+            truncation=True,
+            max_length=self.max_length - 1,
+        )["input_ids"]
+
+        bos = self.tokenizer.bos_token_id
+        eos = self.tokenizer.eos_token_id
```
Comment (aaishwarymishra) [inline trainite/datasets/wikitext.py]: Transform for hugging face datasets should not be responsible for the special token addition, tokenizer should do it.

File: trainite/config/datasets.py [hunk 2]
```diff
@@ -142,3 +142,37 @@ class HuggingFaceDataConfig(DataWithAutoSplit):
...
+
+
+class WikiTextDatasetConfig(HuggingFaceDatasetConfig):
+    path: Literal["Salesforce/wikitext"] = "Salesforce/wikitext"
+    name: Literal[
+        "wikitext-103-raw-v1",
+        "wikitext-103-v1",
+        "wikitext-2-raw-v1",
+        "wikitext-2-v1",
+    ] = "wikitext-2-raw-v1"
```
Comment (aaishwarymishra) [inline trainite/config/datasets.py]: Does this dataset comes with the train, test or validation split on hugging face?

File: trainite/datasets/wikitext.py (hunk 1 above)
Comment (esthermichelle) [reply trainite/datasets/wikitext.py]: You're right, I'll update the transform to let the tokenizer handle the special tokens instead of adding BOS/EOS manually

File: trainite/config/datasets.py (hunk 2 above)
Comment (esthermichelle) [reply trainite/config/datasets.py]: Yes, the wikitext configurations provide train, validation, and test splits on Hugging Face, and the current configuration uses these splits through the dataset loader.

**Comment by esthermichelle** [comment]
I actually wrote the initial implementation myself, and then used AI to help me review and improve parts of it based on the existing architecture and patterns. Thank you for the feedback.

File: trainite/config/datasets.py (hunk 2 above)
Comment (aaishwarymishra) [reply trainite/config/datasets.py]: okay, not really, we have 2 configs for the data config, one in which we split from a dataset like used here, the other where we define separate dataset config for train val etc

Comment (esthermichelle) [reply trainite/config/datasets.py]: Thanks for the clarification, I understand the distinction now. I updated WikiTextDataConfig to use DataConfigBase with separate train, validation, and test SplitConfig entries, since WikiText provides these splits directly

File: tests/datasets/wikitext_test.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+class FakeTokenizer:
+    def __call__(
+        self,
+        text,
+        add_special_tokens=True,
+        truncation=True,
+        max_length=128,
+    ):
+        assert add_special_tokens is True
+        assert truncation is True
```
Comment (aaishwarymishra) [inline tests/datasets/wikitext_test.py]: why we asserting here?

File: tests/datasets/wikitext_test.py
Comment (aaishwarymishra) [inline tests/datasets/wikitext_test.py]: why are the tests scope limited to transform? we should also test the dataset itself

File: trainite/config/datasets.py [hunk 4]
```diff
@@ -142,3 +149,65 @@ class HuggingFaceDataConfig(DataWithAutoSplit):
...
```
Comment (aaishwarymishra) [inline trainite/config/datasets.py]: I dont like this using literals as types, I think many configs use them @TahaZahid05 should we maybe remove them to make code cleaner?

File: trainite/config/registry.py [hunk 5]
```diff
@@ -88,6 +88,17 @@ def dataset_config_cls(self) -> type[BaseModel]:
         preprocessor_spec_name="gpt2",
         dependencies=["datasets"],
     ),
+    "wikitext": DatasetSpec(
+        name="wikitext",
+        implementation_path=Path("trainite/datasets/wikitext.py"),
+        config_cls_path="trainite.config.datasets.WikiTextDataConfig",
+        dataset_config_cls_path="trainite.config.datasets.WikiTextDatasetConfig",
```
Comment (aaishwarymishra) [inline trainite/config/registry.py:95]: @TahaZahid05 we use path for few paths and str for some, I think it could be a separate issue to make them consistent.

File: trainite/datasets/wikitext.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        token_ids = tokenized["input_ids"]
+        attention_mask = tokenized["attention_mask"]
+
+        input_ids = torch.tensor(
+            token_ids[:-1],
+            dtype=torch.long,
+        )
+        labels = torch.tensor(
+            token_ids[1:],
+            dtype=torch.long,
```
Comment (aaishwarymishra) [inline trainite/datasets/wikitext.py:59]: why are we shifting the labels shouldn't this be done by collate_fn, I don't exactly remember how other transform of other dataset behave so i might be wrong :)

File: tests/datasets/wikitext_test.py (hunk 3 above)
Comment (esthermichelle) [reply tests/datasets/wikitext_test.py]: These assertions verify that WikiTextTransform passes the expected arguments to the tokenizer, I can switch this to mock.assert_called_with for consistency if preferred

File: trainite/datasets/wikitext.py (hunk 6 above)
Comment (esthermichelle) [reply trainite/datasets/wikitext.py:59]: The shift is done in the transform following the existing PromptCompletionTransform pattern, while CausalLMCollateFn handles batching and padding, I checked the existing transforms and kept WikiText consistent with that design

File: tests/datasets/wikitext_test.py (hunk 3 above)
Comment (aaishwarymishra) [reply tests/datasets/wikitext_test.py]: Do we even need them?

File: trainite/datasets/wikitext.py (hunk 6 above)
Comment (aaishwarymishra) [reply trainite/datasets/wikitext.py:59]: @vfdev-5 should the dataset do the shifting ? Now I think about it it should be done by collate_fn tied to model, in that case we would need to have multiple collate_fn for transformer itself, like causal collate_fn

File: tests/datasets/wikitext_test.py (hunk 3 above)
Comment (esthermichelle) [reply tests/datasets/wikitext_test.py]: Yeah, I'd keep them , I thought they were useful for catching regressions in add_special_tokens and truncation, but I don't mind removing them if you think they're not needed

File: trainite/datasets/wikitext.py (hunk 6 above)
Comment (vfdev-5) [reply trainite/datasets/wikitext.py:59]: This is a datapoint transform, so it is fine like that.

File: trainite/config/datasets.py (hunk 4 above)
Comment (vfdev-5) [reply trainite/config/datasets.py]: yes, put str here and in the comment a link to the HF dataset, URL, to know the possible options

**Comment by vfdev-5** [comment]
Checking the PR, I do not understand how to create this wikitext dataset programmatically without the config.

**Comment by aaishwarymishra** [comment]
> Checking the PR, I do not understand how to create this wikitext dataset programmatically without the config.

I am afraid, I did not get what you meant. For the dataset, the approach is to simply use the `dataset.load_dataset` as the `__target__` and pass the parameter like path are then passed to the load_dataset. For the pr in question they simply used the generic  HuggingFaceDatsetConfig  as the parent.

File: trainite/config/datasets.py (hunk 4 above)
Comment (esthermichelle) [reply trainite/config/datasets.py]: I'll update it to str and add the Hugging Face dataset URL in the comment, Thank you for the feedback.

File: trainite/config/registry.py [hunk 7]
```diff
@@ -88,6 +88,17 @@ def dataset_config_cls(self) -> type[BaseModel]:
...
+        implementation_symbol="WikiTextTransform",
+        builder_symbol="datasets.load_dataset",
+        readme_template_path=Path("trainite/templates/components/datasets/wikitext.md"),
+        preprocessor_spec_name="gpt2",
+        dependencies=["datasets"],
+    ),
```
[the remark this answers is not in the document]
Comment (aaishwarymishra) [reply trainite/config/registry.py:101]: I think its outdated now,
Edit:that pr to remove the hardcoded parameters on cli generation test is not merged yet, for the formalityI think we can add it for now.

File: tests/datasets/wikitext_test.py (hunk 3 above)
Comment (aaishwarymishra) [reply tests/datasets/wikitext_test.py]: we dont need it if you pass correct parameters in the test file

File: trainite/config/registry.py (hunk 5 above)
Comment (TahaZahid05) [reply trainite/config/registry.py:95]: sure i'll open it.

**Comment by esthermichelle** [comment]
Got it, I'll remove the unnecessary assertions and add the wikitext case to the CLI generation test for now, Thanks for the clarification!

File: trainite/config/registry.py (hunk 5 above)
Comment (TahaZahid05) [reply trainite/config/registry.py:95]: @aaishwarymishra Actually, we have to use Path and str depending on whether we want the template file or the import target. So the use of Path and str is correct. We have to standardize the naming. For instance, when using strs we can have `config_cls_target` so we can differentiate between target and paths. Does that make more sense?

**Comment by esthermichelle** [comment]
@aaishwarymishra could you please review the latest changes, I’ve addressed the review feedback, Thanks for your patience!

File: trainite/config/registry.py (hunk 5 above)
Comment (aaishwarymishra) [reply trainite/config/registry.py:95]: yeah, it does

Comment (esthermichelle) [reply trainite/config/registry.py:95]: Just to clarify, should I update the naming in this PR? @aaishwarymishra

Comment (aaishwarymishra) [reply trainite/config/registry.py:95]: Not really if should be a separate pr, I will look after this pr Tommorow I hope its fine:)

File: tests/cli_test.py [hunk 8]
```diff
@@ -91,15 +92,23 @@ def test_init_generates_valid_project(models: list[str], dataset: str, trainer:
...
             assert generated_config["data"]["dataset"]["_target_"] == "datasets.load_dataset"
-            assert generated_config["preprocessor"]["_target_"] == ("preprocessors.gpt2_tokenizer.load_gpt2_tokenizer")
+            assert generated_config["preprocessor"]["_target_"] == "preprocessors.gpt2_tokenizer.load_gpt2_tokenizer"
             generated_pyproject = (project_dir / "pyproject.toml").read_text()
             assert "datasets" in generated_pyproject
             assert "transformers" in generated_pyproject
+
         else:
             assert generated_config["data"]["dataset"]["_target_"].startswith("dataset_impl.")
-        assert generated_config["data"]["transform"]["_target_"].startswith("dataset_impl.")
```
Comment (aaishwarymishra) [inline tests/cli_test.py:109]: why they are repeated?

File: tests/cli_test.py [hunk 9]
```diff
@@ -91,15 +92,23 @@ def test_init_generates_valid_project(models: list[str], dataset: str, trainer:
             if filename is not None:
                 py_compile.compile(str(project_dir / filename), doraise=True)
```
Comment (aaishwarymishra) [inline tests/cli_test.py:110]: this code is almost similar to the hugging face if statement, I think the hugging face datasets should have a common if block, I thing is how to identify them without having to write big if condition to check every hugging face sataset

**Comment by esthermichelle** [comment]
Thanks for pointing this out, I’ll refactor the test to use a common condition for hugging face-backed datasets instead of checking each dataset by name, I’ll also handle the different data split structures and keep the dataset\transform assertions together to avoid the repeated-looking logic, I’ll update the test accordingly

File: tests/cli_test.py
Comment (aaishwarymishra) [inline tests/cli_test.py]: looks good, but why there are some unnecessary diffs

**Comment by aaishwarymishra** [comment]
Hi, other code looks fine to me, these are some final changes, can you replace the literals with str in pydantic classes, a recent pr got merged that makes all the  literals to str.

And pls take a fresh pull of this branch , I resolved a merge conflict.

And address one more comment I left, and I think this PR can be merged.

Thank you for your help :)

**Comment by esthermichelle** [comment]
Done @aaishwarymishra

**Review by aaishwarymishra — APPROVED** [review]

**Comment by aaishwarymishra** [comment]
Thanks
